ci-bench/src/main.rs

```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    Ok(())
+}
+
+/// Returns all benchmarks
+fn get_benchmarks() -> anyhow::Result<Vec<Benchmark>> {
+    let mut benchmarks = Vec::new();
+    for &param in ALL_BENCHMARK_PARAMS {
+        add_benchmark_group(&mut benchmarks, param);
+    }
+
+    validate_benchmarks(&benchmarks)?;
+    Ok(benchmarks)
+}
+
+/// The benchmark params to use for each group of benchmarks
+static ALL_BENCHMARK_PARAMS: &[BenchmarkParams] = &[
+    BenchmarkParams::new(
+        KeyType::Rsa,
+        rustls::cipher_suite::TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+        &rustls::version::TLS12,
+        "1.2_rsa_aes",
+    ),
+    BenchmarkParams::new(
+        KeyType::Rsa,
+        rustls::cipher_suite::TLS13_AES_128_GCM_SHA256,
+        &rustls::version::TLS13,
+        "1.3_rsa_aes",
+    ),
+    BenchmarkParams::new(
+        KeyType::Ecdsa,
+        rustls::cipher_suite::TLS13_AES_128_GCM_SHA256,
+        &rustls::version::TLS13,
+        "1.3_ecdsa_aes",
+    ),
+    BenchmarkParams::new(
+        KeyType::Rsa,
+        rustls::cipher_suite::TLS13_CHACHA20_POLY1305_SHA256,
+        &rustls::version::TLS13,
+        "1.3_rsa_chacha",
+    ),
+    BenchmarkParams::new(
+        KeyType::Ecdsa,
+        rustls::cipher_suite::TLS13_CHACHA20_POLY1305_SHA256,
+        &rustls::version::TLS13,
+        "1.3_ecdsa_chacha",
+    ),
+];
+
+/// Adds a group of benchmarks for the specified parameters
+///
+/// The benchmarks in the group are:
+///
+/// - Handshake without resumption
+/// - Handshake with session id resumption
+/// - Handshake with ticket resumption
+/// - Transfer a 1MB data stream from the server to the client
+fn add_benchmark_group(benchmarks: &mut Vec<Benchmark>, params: BenchmarkParams) {
+    let params_label = params.label;
+
+    // Create handshake benchmarks for all resumption kinds
+    for &resumption_param in ResumptionKind::ALL {
+        let handshake_bench = Benchmark::new(
+            format!("handshake_{}_{params_label}", resumption_param.label()),
+            BenchmarkKind::Handshake(resumption_param),
+            params,
+        );
+
+        let handshake_bench = if resumption_param != ResumptionKind::No {
+            // Since resumed handshakes include a first non-resumed handshake, we need to subtract
+            // the non-resumed handshake's instructions
+            handshake_bench
+                .exclude_setup_instructions(format!("handshake_no_resume_{params_label}"))
+        } else {
+            handshake_bench
+        };
+
+        benchmarks.push(handshake_bench);
+    }
+
+    // Benchmark data transfer
+    benchmarks.push(
+        Benchmark::new(
+            format!("transfer_no_resume_{params_label}"),
+            BenchmarkKind::Transfer,
+            params,
+        )
+        .exclude_setup_instructions(format!("handshake_no_resume_{params_label}")),
+    );
+}
+
+/// Runs a benchmark for the client-side of the connection
+fn run_bench_for_client(
```

Comment:
I think I still maintain that this would be better if you had a `Handshake` trait or similar that takes the `new_client()`, `handshake_client_side()` (maybe these can be combined?) and `handshake_client()` side out, so there's no need to deduplicate `run_bench_for_client()` and `run_bench_for_server()`.

Reply:
I'll have a look into it

Reply:
Just pushed a commit which maces `run_bench` generic based on a `BenchStepper` trait. Does that look like what you had in mind?

Reply:
It does seem like an improvement. I'd probably factor out the 3 common fields of the `Stepper` types into a separate type and also attach the `make_new_config()` types as methods to the `{Client,Server}SideStepper` types.